### PR TITLE
Replace StochasticFunctions v2

### PR DESCRIPTION
**Pull request by colesbury**
This removes the StochasticFunctions for bernoulli, multinomial, and
normal and replaces them with classes in the torch.distributions
package. Each distribution supports the differentiable log_prob function
that returns the log of the pdf/pmf of the samples.

The current StochasticFunction implementation has a few problems: it can
be painful to use when there are multiple stochastic outputs which need
to be back-propagated through. It also requires that we store grad_fns
on Variables that have requires_grad=False in order to find stochastic
nodes.

### CODE REVIEW

**Comment by colesbury** [comment]
cc @soumith @apaszke @Kaixhin

File: torch/distributions.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        self.mean = mean
+        self.std = std
+
+    def sample(self):
+        return torch.normal(self.mean, self.std)
+
+    def log_prob(self, value):
+        # compute the variance
+        var = (self.std ** 2)
+        return -((value - self.mean) ** 2) / (2 * var) - (self.std).log() - math.log(math.sqrt(2 * math.pi))
```
Comment (apaszke) [inline torch/distributions.py]: `(self.std).log()` won't work if `self.std` is a `float`

**Review by apaszke — COMMENTED** [review]
LGTM. It feels a bit weird that the shape returned by `sample` depends on the constructor arguments. You sometimes want to get more than a single sample, and if you give both `mean` and `std` to `Normal` constructor, there's no shape information. I think `sample` should be able to accept extra args. It doesn't make sense to pretend that you can freely swap one distribution for another.

**Comment by colesbury** [comment]
We'll want a `def sample_n(num_samples)` function that samples multiple times. I didn't include it because I wanted to keep this minimal.

It'll also be better once we have better scalar support.

TensorFlow calls the shape determined by the parameters the `batch shape`. `sample()` returns a tensor with the batch shape. `sample_n(N)` returns something `N` x `batch_shape`.

For Bernoulli and Normal the batch shape is just the shape of the constructor arguments.
For Mutlinomial, the probs is `batch_shape` x `K`.

This means that the returned shapes follow a consistent pattern. We want `sample_n` to be a separate function from `sample` because `1 x N` is a different shape from `N`.

File: torch/distributions.py (hunk 1 above)
Comment (colesbury) [reply torch/distributions.py]: I'm going to require that mean and std are Tensor/Variables for now. Once we have better scalar support we can allow numbers

**Comment by apaszke** [comment]
Alright, `sample_n` sounds good to me.

File: torch/distributions.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+    Args:
+        mean (Tensor or Variable): mean of the distribution
+        std (Tensor or Variable): standard deviation of the distribution
+    """
+    def __init__(self, mean, std):
+        if not torch.is_tensor(mean) and not isinstance(mean, Variable):
+            raise TypeError('mean must be a Tensor or Variable')
+        if not torch.is_tensor(std) and not isinstance(std, Variable):
+            raise TypeError('std must be a Tensor or Variable')
```
Comment (apaszke) [inline torch/distributions.py]: I think it makes sense to allow one of them to be a number, just forbid both. E.g. you want your model to generate means and you want to sample with a fixed variance

**Review by apaszke — APPROVED** [review]
LGTM. I would only relax the constraints on `Normal`

**Comment by hughperkins** [comment]
To save other people hunting around, looks like these work something like:

```
m = torch.distributions.Multinomial(probs)
a = m.sample()
loss = - (m.log_prob(a) * reward).sum()
loss.backward()
```

(where before we'd have done:
```
a = torch.multinomial(probs)
a.reinforce(reward)
autograd.backward([a], [None])
```

**Comment by Kaixhin** [comment]
@hughperkins That is a super useful comment - I'll submit a PR to fix this in the RL examples in `pytorch/examples`. If you can think of anywhere else we need to update this, please go ahead.

**Comment by colesbury** [comment]
Yeah, we should also update Variable.reinforce to raise an exception with a hint about what you need to do.

**Comment by hughperkins** [comment]
cool. actually, I think I'm restructuring it slightly:
```
m = torch.distributions.Multinomial(probs)
a = m.sample()
eligibility = m.log_prob(a)
a = a.data

# (and then later)

loss = - (eligibility * reward).sum()
loss.backward()
```
Thoughts?

**Comment by Kaixhin** [comment]
The term "eligibility" may be confusing. I guess it comes from the original work on REINFORCE, but "eligibility traces" are by far the most prevalent use of the word "eligibility" in RL. I would stick with `m.log_prob(a)` as it is quite short and self-descriptive, unless you have a strong reason for adding in the extra term?

**Comment by hughperkins** [comment]
well, typically, I would think that the policy where we generate the action and the partial loss term could be well separated from where we actually act on the action, and apply the loss/reward?

For example, policy could be way up here:

https://github.com/ASAPPinc/emergent_comms_negotiation/blob/fb50c3979f240d8eff07db5aef08b4f4f59a3309/ecn.py#L148-L154
```
    def forward(self, x, eps=1e-8):
        x = self.h1(x)
        x = F.sigmoid(x)
        m = torch.distributions.Bernoulli(x)
        a = m.sample()
        eligibility = m.log_prob(a)
        a = a.data
        x = x + eps
        entropy = - (x * x.log()).sum(1).sum()
        return eligibility, a, entropy
```

but action isnt acted on till much later, eg in this case, the action is to finish the game, and calculate the reward, if the terminate action was true:
https://github.com/ASAPPinc/emergent_comms_negotiation/blob/fb50c3979f240d8eff07db5aef08b4f4f59a3309/ecn.py#L358
```
        # calcualate rewards for any that just finished
        reward_eligible_mask = term_a.view(batch_size).clone().byte()
```
... which is in a totally different part of the code

... and then later on we calculate hte loss and do backprop, again in a totally different part of the code, eg
https://github.com/ASAPPinc/emergent_comms_negotiation/blob/fb50c3979f240d8eff07db5aef08b4f4f59a3309/ecn.py#L519
```
loss_by_agent[agent] -= (action * Variable(alive_rewards[:, agent].contiguous().view(_batch_size, 1))).sum()
```

So... if we multiply out the `m.log_prop(a)` in the policy, we can pass that bit around, without needing to take the `a` with it. Similarly, since we're never going to backprop through `a`, we may as well strip the `Variable` wrapper, and pass `a` around as a tensor.

**Comment by Kaixhin** [comment]
OK I think I see where you're coming from. Yes, in general these steps are quite separated. In the [docs](http://pytorch.org/docs/master/distributions.html), this separation would be artificial and I would favour keeping it as it is to closely conform to the mathematical equation for the REINFORCE rule. For the examples we can actually mimic what is done in practice, as you've demonstrated above. How does that sound? I can send in a PR for the docs here (in the main repo) with the REINFORCE rule in MathJax.

**Comment by hughperkins** [comment]
Cool, sounds good :)

**Comment by hughperkins** [comment]
(PS Opinion: personally, I like the new torch.distributions.xxx interface; I find it very easy to understand, personally; very close to the equations we actually see in REINFORCE etc papers :+1: ).
